Test jvm (#1)
LucasDblt/QBDL: examples/tools/android_native_run/readme.md

# Support for Android native libraries to QBDL

The goal of this project was to launch a Java Virtual Machine which is able to support the unknown android symbols. 
For our application, we are using some code from android studio. The c++ program will then start the JVM in which
the Java code from android studio will be able to compil thanks to the imported libraries.

## How to use it :

### Java classes

To compile the code, you will need to know the paths to the Java classes. Then create a .txt document in which you can gather all these paths. To do so, you can use this command :

> `find ~/.gradle/caches -name "*-runtime.jar" > dependencies.txt`

Reviewer: When putting code in a separate paragraph you can use "```" to open/end the code block.

Author: Done

> It will find all the java dependencies ending with "runtime.jar" which are necessary for the code to be well executed. Other files are required although. You have to include these two paths into the same file :

Reviewer: ```suggestion
It will find all the java dependencies ending with "runtime.jar" which are necessary for the code to be well executed. Other files are required also. You have to include these two paths into the same file :
```
`although` -> `also` or `though`, both could work in that context.

Author: Done

> > The Android runtime, which you can normally find in your home directory followed by :
> `/Android/Sdk/platforms/android-32/android.jar`

Reviewer: Using a `>` here makes things look a bit weird when rendering the markdown, you can probably use a list here:
```suggestion
  - The Android runtime, which you can normally find in your android's SDK installation followed by `/platforms/android-32/android.jar`.
  - The directory where the Java classes using the native libraries are compiled; on my setup it was here:
`~/dev/example-android-native-app/app/build/intermediates/javac/debug/classes`
```

Author: Done

> The repertory where the Java classes using the native libraries are compiled is also needed :
`/dev/example-android-native-app/app/build/intermediates/javac/debug/classes`

> ### LD_library_path

Reviewer: ```suggestion
### LD_LIBRARY_PATH
```

Author: Done

> When the code called by the JNI is loading, it is loading itself the native library. Then the right LD_library_path has to be given to it.

Reviewer: ```suggestion
When the code called by the JNI is loading, it is loading itself the native library. Therefore the right LD_LIBRARY_PATH has to set so that it finds the necessary libraries.
```

Author: Done

> You can now export the LD_library_path using `export LD_LIBRARY_PATH=path/to/libc.so:/path/to/libmyapplication.so`

Reviewer: ```suggestion
You can now export the LD_LIBRARY_PATH using `export LD_LIBRARY_PATH=path/to/libc.so:/path/to/libmyapplication.so`
```

Author: Done

### Options to execute

Now after compilation, you will have to execute the code with these options :

`-d` path to the .txt with the java dependencies

`-l` path to the .txt with the other libraries to add

`-i` ld_path, we are using here "libmyapplication.so"

> ## Explanation on the different sections of the code

Reviewer: Using `sections` here is slightly confusing because we are manipulating binaries and the word `section` has a very specific meaning in that context.
I could recommend using simply `the different parts of the code`.

Author: Done

> > __Options section__

Reviewer: `>` is generally used for quotes; if you want to denote a subsection you can use `####` or something like `**Options**`.

Author: Done

```
jvmopt[0].optionString = const_cast<char *>(opt.class_path.c_str());
jvmopt[1].optionString = const_cast<char *>(opt.library_path.c_str());
```
Now the JVM has all the needed informations to start and to call the JNI !

> __Myfunctions section__

Here is grouped all the unknown functions which we are going to replace. When a symbol is not recognized (not present in the basic libc), the program will look into this library to find a remplacement function.

> __Other_libraries section__

This section works a bit like the classpath section but with the libraries. It create a vector from the informations in Other_libraries.txt.

> __Main section__

The most important part since this section will reunify every part for the program to work.

The first `namespace{...}` part is using the Other_libraries section to open and load every of these libraries.

The `int main(int argc, char **argv) {...}` part is firstly calling the Option section to use or redistibute them. Then calling the Java functions add and nativeMul from the JNI. This second part is just an application of the project but the program can be modified to be use with any other functions created with android studio.

> ## Some unnecessary symbols 

Reviewer: ```suggestion
## Removing unused libraries
```

Author: Done
